lib/src/kvs/sled/mod.rs

```diff
@@ -0,0 +1,782 @@
+#![cfg(feature = "kv-sled")]
+
+use crate::err::Error;
+use crate::kvs::Key;
+use crate::kvs::Val;
+use sled::{Db, Iter};
+use std::collections::btree_map::Range as BTreeMapRange;
+use std::collections::{BTreeMap, HashSet};
+use std::ops::Range;
+
+pub struct Datastore {
+	db: Db,
+}
+
+pub struct Transaction {
+	// Is the transaction complete?
+	ok: bool,
+	// Is the transaction read+write?
+	rw: bool,
+	// The DB this translation is related to
+	db: Db,
+	// Key/Value pair to set
+	set: Option<BTreeMap<Key, Val>>,
+	// Keys to delete
+	del: Option<HashSet<Key>>,
+}
```

Comment:
Is there a good reason for `Option` here?

Note that empty `BTreeMap`'s and `HashSet`'s don't allocate memory, and using `Option` seems to introduce a lot of boilerplate code later on.
- easier to add the first item
- can remove many `if let Some(...) = ... { }`
- `set = None` -> `set.clear()`

Reply:
@finnbear I agree that it is a trade-off. A typical trade-off is using a Set rather than an Array for very small arrays. That said, here I believe we have a real opportunity. So, I wrote two tests to gather metrics.

These tests are not perfect, but I believe they are fair comparisons of both option performances.

The test `test_transaction_hashset_vs_option_rationale`, shows that `Option::is_none` is more than 4 times faster than `HashSet::is_empty()`. 

```
set: 1050
option: 233
option_set: 1008
set / option : 4.51
set / option_map : 1.04
```

The test `test_transaction_btreemap_vs_option_rationale` shows an improvement by a factor of more than 3.

```
map: 1294
option: 358
option_map: 1291
map / option : 3.61
map / option_map : 1.00
```

The rationale was that we will probably have 3 kinds of transactions.
- only additions, 
- only deletions,
- mixed additions and deletions.

In the end, this is a gain in time execution, but also in space.

I would also mention that BTreeMap is using a similar approach with its [Root node](https://github.com/rust-lang/rust/blob/a102dc806da3bc9c59b3594368a14e7d2632bf9c/library/alloc/src/collections/btree/map.rs#L177).

Reply:
> In the end, this is a gain in time execution, but also in space.

**I agree time and space are both important, and could serve to justify a given design. Since you do have reasons for `Option`, feel free to mark this discussion as resolved**. However...

As for space, I already mentioned how empty collections don't allocate on the heap but I should also note that `enum`'s such as `Option` need as much memory as their *largest* variant with the possible addition of a discriminant.

```

use std::collections::{BTreeMap, HashSet};
pub type Key = Vec<u8>;
pub type Val = Vec<u8>;

fn main() {
    type Map = BTreeMap<Key, Val>;
    type Set = HashSet<Key>;
    println!("map={} opt map={}", std::mem::size_of::<Map>(), std::mem::size_of::<Option<Map>>());
    println!("set={} opt set={}", std::mem::size_of::<Set>(), std::mem::size_of::<Option<Set>>());
}

map=24 opt map=32
set=48 opt set=48
```
In the end, `Option<BTreeMap<Key, Val>>` (regardless if `Some` or `None`) will consume 33% more memory than `BTreeMap<Key, Val>`). `HashSet<Key>`'s memory use is unaffected by `Option`.

As for execution time, I was a bit surprised that `Option` would be faster. I modified your benchmarking code such that it:
1. uses `std::hint::black_box` to defeat over-optimization
2. runs `is_empty`/`is_none`/`is_some` more than once on each new map
3. ran in `release` mode (inlining should benefit `is_empty` due to deep function call indirection)

<details>
<summary>New test code</summary>

```rust
#![feature(bench_black_box)]

use std::collections::{BTreeMap, HashSet};
pub type Key = Vec<u8>;
pub type Val = Vec<u8>;

fn main() {
    type Map = BTreeMap<Key, Val>;
    type Set = HashSet<Key>;
    println!("map={} opt map={}", std::mem::size_of::<Map>(), std::mem::size_of::<Option<Map>>());
    println!("set={} vs. opt set={}", std::mem::size_of::<Set>(), std::mem::size_of::<Option<Set>>());

    test_transaction_btreemap_vs_option_rationale();
    test_transaction_hashset_vs_option_rationale();
}

use std::time::SystemTime;
const N: usize = 100000;
const M: usize = 10;

fn test_transaction_btreemap_vs_option_rationale() {
    // Evaluating BTreeMap creation + BTreeMap::is_empty()
    let time = SystemTime::now();
    for _ in 0..N {
        let map = BTreeMap::<Key, Val>::new();
        for _ in 0..M {
            assert!(std::hint::black_box(&map).is_empty());
        }
    }
    let map_time = time.elapsed().unwrap();

    // Evaluating Option creation + Option::is_none()
    let time = SystemTime::now();
    for _ in 0..N {
        let map: Option<BTreeMap<Key, Val>> = None;
        for _ in 0..M {
            assert!(std::hint::black_box(&map).is_none());
        }
    }
    let option_none_time = time.elapsed().unwrap();

    // Evaluating Map encapsulated in an Option + Option::is_none()
    let time = SystemTime::now();
    for _ in 0..N {
        let map = Some(BTreeMap::<Key, Val>::new());
        for _ in 0..M {
            assert!(std::hint::black_box(&map).is_some());
        }
    }
    let option_some_time = time.elapsed().unwrap();

    let map_vs_option_none = map_time.as_micros() as f64 / option_none_time.as_micros() as f64;
    let map_vs_option_some = map_time.as_micros() as f64 / option_some_time.as_micros() as f64;

    println!("map: {}", map_time.as_micros());
    println!("option_none: {}", option_none_time.as_micros());
    println!("option_some: {}", option_some_time.as_micros());
    println!("map / option_none : {:.2}", map_vs_option_none);
    println!("map / option_some : {:.2}", map_vs_option_some);
}

fn test_transaction_hashset_vs_option_rationale() {
    // Evaluating BTreeMap creation + HashSet::is_empty()
    let time = SystemTime::now();
    for _ in 0..N {
        let set = HashSet::<Key>::new();
        for _ in 0..M {
            assert!(std::hint::black_box(&set).is_empty());
        }
    }
    let set_time = time.elapsed().unwrap();

    // Evaluating Option creation + Option::is_none()
    let time = SystemTime::now();
    for _ in 0..N {
        let set: Option<HashSet<Key>> = None;
        for _ in 0..M {
            assert!(std::hint::black_box(&set).is_none());
        }
    }
    let option_none_time = time.elapsed().unwrap();

    // Evaluating Map encapsulated in an Option + Option::is_none()
    let time = SystemTime::now();
    for _ in 0..N {
        let set = Some(HashSet::<Key>::new());
        for _ in 0..M {
            assert!(std::hint::black_box(&set).is_some());
        }
    }
    let option_some_time = time.elapsed().unwrap();

    let set_vs_option_none = set_time.as_micros() as f64 / option_none_time.as_micros() as f64;
    let set_vs_option_some = set_time.as_micros() as f64 / option_some_time.as_micros() as f64;

    println!("set: {}", set_time.as_micros());
    println!("option_none: {}", option_none_time.as_micros());
    println!("option_some: {}", option_some_time.as_micros());
    println!("set / option_none : {:.2}", set_vs_option_none);
    println!("set / option_some : {:.2}", set_vs_option_some);
}
```
</details>

I got the following numbers (assuming 10 emptiness checks per new map):
```
map: 4156
option_none: 2944
option_some: 7770
map / option_none : 1.41
map / option_some : 0.53
set: 5746
option_none: 3211
option_some: 5841
set / option_none : 1.79
set / option_some : 0.98
```

As the number of emptiness checks approaches infinity, the performance difference approaches zero. This implies that the main overhead saved by `Option` is creating new `BTreeMap`'s and `BTreeSet`'s. Apparently their constructors take some time, regardless if they don't need to allocate on the heap.

Feel free to do whatever you want with this information! (including closing this discussion as resolved)

Reply:
@finnbear Thanks for the information there, very appreciated! Benchmarking is for sure a much broader topic and should be handled in a dedicated way with the proper tools and the proper approach.  That said, I have updated the test, introducing randomized states. When running the test in release mode I have the same results, pointing out that the constructor is taking time, `option::is_none` and `both hashset::is_empty` and `btreemap::is_empty` providing pretty similar performances. Avoiding the construction time (on transactions that either don't require deletion or don't require insertion) was my initial intent. So far so good!

Reply:
@tobiemh The next step is testing this implementation.  I have tried `cargo test --features storage-sled` which does not seem to trigger any test.
Tried also, in the lib directory, `cargo test --features kv-sled` ends up with errors (also without the feature):

```bash
...
test sql::value::array::tests::array_none ... FAILED
test sql::value::array::tests::array_path ... FAILED
test sql::value::clear::tests::clear_value ... FAILED
test sql::statements::select::tests::select_statement_table_thing ... ok
test sql::value::compare::tests::compare_basic_missing_left ... ok
test sql::value::compare::tests::compare_basic_missing_right ... ok
test sql::value::compare::tests::compare_basic ... ok
test sql::value::decrement::tests::dec_array_array ... FAILED
test sql::value::decrement::tests::dec_array_number ... FAILED
test sql::value::decrement::tests::dec_array_value ... FAILED
test sql::value::decrement::tests::dec_none ... FAILED
...
```

What is the current proper way for testing a database engine?

Reply:
I used this command (extracted from the GH action): 

```
(&>/dev/null cargo run -- start --log trace --user root --pass root memory &)
          cargo test --workspace --features protocol-ws,protocol-http,kv-rocksdb,kv-sled
```